### PR TITLE
775: Support Entities from repeats 

### LINKED ISSUE #775: Support Entities from repeats
Specification: https://forum.getodk.org/t/create-and-update-entities-from-repeats/55163

The writeup above says only one row will be initially allowed on the `entities` sheet at a time. That means a form could either be configured to create/update a single entity OR multiple Entities from a repeat but not both (e.g. not a household Entity and multiple Member Entities). There's a chance we'll expand initial scope to include multiple Entity declarations but if so, let's track it in a follow-on issue.

### PR DESCRIPTION
**Pull request by lindsay-stevens**
Closes #775
Addresses question name issue in #504

#### Why is this the best possible solution? Were any other approaches considered?

The spec involves putting a "meta" block inside each repeat as well as at the top-level survey/main-instance node. This conflicts with existing name validation. However it seems that the name rules in pyxform are inconsistent and too restrictive. To enable the new entities feature, the modifications are made to name validation such that the new rules would be as follows.

- all names are case-sensitive (survey, repeat, group, question, choice list, choice, entity)
  - if SAS or SPSS will be used to analyse data, it's recommended to make names case-insensitive unique within their context (e.g. avoid "q1" and "Q1").
- repeat names must be unique anywhere in the survey (at any level of nesting), and cannot be the same as the survey instance name (default "data").
  - if the XLSForm was converted separately with pyxform (not via xlsform-online or Central) then the survey instance name is the file name.
- external instances ("xml-external", "csv-external") must be unique anywhere in the survey (at any level of nesting).
- question and group names must be unique within their context i.e. nearest parent group or repeat, or the survey if not inside a group or repeat.
  - if pyxform reference variables "${}" are used, referenced name (question, group, repeat) must unique anywhere in the survey
    - if this is not possible then the pyxform reference can be replaced by a manually written XPath expression.
  - a loop produces a group in the loop's context, and so loop names follows the same rule for group names.
  - if the (deprecated) "flat" setting is active, question names must be unique anywhere in the survey.
  - pyxform may generate the following questions or groups, so these names are reserved:
    - "meta" which is used for survey metadata such as instanceID and entity configuration
      - member questions include: "audit", "entity", "instanceID", "instanceName"
    - "generated_note_name_{row_number}" if a note does not have a name
    - within a "loop", a group named with the choice name will contain a copy of each question or group inside the loop
      - loops cannot contain repeats since this would produce duplicate repeat names
- choice list names must be unique anywhere in the survey.
  - choice lists are internal instances so the names must not be the same as other instances from:
    - the file name (excluding extension) from any "select_one_from_file" or "select_multiple_from_file" in the survey
    - the first argument of any "pulldata" calls that refer to an instance that doesn't already exist
- choice names must be unique within the choice list
  - if the "allow_choice_duplicates" setting is active ("y") then choice names can be duplicated within each list
  - duplicate choices and loops are not allowed, because this would generate duplicate groups in the same context

The main modification is to allow a group to be unique only within it's context, rather than anywhere in the survey (as will be the case for the `meta` block). The main constraint seems to be that Central will produce export files for the top-level survey, and any repeats; any group names in such files are kebab'd together within the context of the survey OR lowest common ancestor repeat e.g. 'g1-g2-question'.

*TODO: commentary on the actual entity repeats changes*

#### What are the regression risks?

I did some git archaeology but I might've missed a critical reason why groups have historically been required to be unique at any depth - maybe Aggregate used to split by group not just repeat? Otherwise hopefully it's relaxing existing rules and improving error messages.

#### Before submitting this PR, please make sure you have:
- [ ] included test cases for core behavior and edge cases in `tests`
- [ ] run `python -m unittest` and verified all tests pass
- [ ] run `ruff format pyxform tests` and `ruff check pyxform tests` to lint code
- [ ] verified that any code or assets from external sources are properly credited in comments

### CODE REVIEW

File: tests/test_fields.py [hunk 1]
```diff
@@ -2,211 +2,513 @@
...
+        | | begin group | q1   | G2    |
+        | | text        | q2   | Q2    |
+        | | end group   |      |       |
         """
-        Ensure that duplicate field names with different cases are not allowed
+        self.assertPyxformXform(
+            md=md,
+            errored=True,
+            error__contains=[
+                "There are more than one survey elements named 'q1' (case-insensitive) in the section named 'test_name'."
```
[the remark this answers is not in the document]
Comment (lindsay-stevens) [reply tests/test_fields.py]: The `test_name` is from the default instance name for PyXFormTestCase, and there's an open ticket #715 to change it to `data` to align with the more common default. When pyxform is used from the command line, the name is the file base name e.g. `my_form` from `my_form.xlsx`; otherwise it's set to `data`. I think I selected `test_name` for tests also to make it easier to do code searches for but in practice I haven't had to search for the instance name very often.

File: pyxform/xls2json.py [hunk 2]
```diff
@@ -60,6 +60,14 @@
 RE_OSM = re.compile(
     r"(?P<osm_command>(" + "|".join(aliases.osm) + r")) (?P<list_name>\S+)"
 )
+INVALID_CONTROL_END = (
+    "[row : {row}] Unmatched 'end {control_type}'. "
```
[the remark this answers is not in the document]
Comment (lindsay-stevens) [reply pyxform/xls2json.py]: Addressed in ba420544 but as noted in the commit message:

- for the name in the error message for unmatched end to be useful, the user would have to have written the name which is optional
- may or may not be helpful to additionally show the name from the previous control with a different type - could be due to a missing row or a typo in the control type.

If it needs further enhancement maybe that could split off to a separate ticket for follow up?

File: tests/entities/test_create_repeat.py [hunk 3]
```diff
@@ -0,0 +1,460 @@
...
+        | | begin_repeat | r1    | R1    |         |
+        | | text         | q1    | Q1    | q1e     |
+        | | end_repeat   |       |       |         |
+
+        | entities |
+        | | list_name | label | repeat |
+        | | e1        | ${q1} | ${r2}  |
+        """
+        self.assertPyxformXform(
+            md=md, errored=True, error__contains=[ep.ENTITY006.format(row=3, value="q1e")]
```
[the remark this answers is not in the document]
Comment (lindsay-stevens) [reply tests/entities/test_create_repeat.py:382]: Yes it's because the save_to issue becomes apparent during the iteration over the survey row data, whereas it's not clear that there's no repeat for the entity repeat reference until the end of that iteration.

**Comment by lognaturel** [comment]
As I was trying this out, I forgot to specify the repeat reference on the `entities` sheet. I put my savetos in the repeat and I still got a toplevel Entity declaration. Could we also show an error when the Entity declaration is toplevel and there's a saveto in a repeat? Or maybe more generally, if there's a saveto for a field for which the reference from the `repeat` column is not an immediate parent, error.

**Comment by lognaturel** [comment]
It looks like all of the references produced in binds are off by one parenting level: https://github.com/lindsay-stevens/pyxform/compare/pyxform-775...lognaturel:pyxform:hm_comments

I'm pretty sure that's because in `EntityDeclaration. _get_bind_node`, the `insert_xpaths` call uses `self` as the context which is the meta block for the Entity. Instead, it should use the attribute or node **in** the meta block. But of course that doesn't exist because it's the very thing we're building at that point! It seemed to work fine when at the toplevel because absolute references were produced anyway but it doesn't work in repeats where that additional parenting level is important.

It doesn't feel directly related to https://github.com/XLSForm/pyxform/pull/777 because `insert_xpaths` is doing the right thing, it's just called with the wrong context. That said, I'll shift my attention back to that PR to see whether we could safely get it merged in before this one. I'm still hoping there's an easy fix for this issue that I'm not seeing. Maybe we can just use a dummy node as context?

I've made a list of must-haves for this PR in the original description so we don't forget anything. Thanks, @lindsay-stevens, lots of tricky stuff here that will hopefully benefit users a ton!

**Comment by lognaturel** [comment]
> the quotes surrounding the "references" symbol are escaped

Whoops 😬

> tests.entities.test_create_repeat.TestEntitiesCreateRepeat.test_other_controls_after__ok

`pyxform` either after this change or always puts `setvalue` in the model. From [the spec](https://getodk.github.io/xforms-spec/#the-odk-new-repeat-event): "Actions triggered by odk-new-repeat must be nested in the repeat form control."

It does look like Collect manages to do the right thing when there's only one repeat but things break when there are multiple and I bet are even worse when there's nesting involved. The ideal fix would be to put the `setvalue` action in the `repeat` and I've verified that does work with Validate/Collect.

**Comment by lindsay-stevens** [comment]
@lognaturel the setvalue issue is addressed in 1e481b9 - overview in commit message but below are some detailed notes on old vs. new. I found actions generation quite confusing but I think it's clearer now and hopefully you agree.

One downside of making actions 1:1 with event types is that trigger expressions can appear verbatim in 2 places for when the trigger is in a repeat, but I would expect that usually these expressions are not tremendously long such that they would dramatically increase XForm file size. I couldn't see any other reason besides brevity to combine the setvalue for that case but please let me know otherwise - it should be possible to allow a sequence of events to combine the actions again.

Also I assumed it was a typo that the entity `setvalue` had until now included the attributes `readonly="true()"` and `type="string"` - these are common attrs for `bind` but I don't know what if anything they mean for an action (they aren't standard XForm action attributes and don't seem to appear in ODK XForm or Entity specs). If they are not typos please let me know and I'll put them back in.

Old output locations, call paths, action/event types, reasons for generation

- output to model:
  - call path: Survey/xml/xml_model/xml_descendent_bindings -> Question/get_setvalue_node_for_dynamic_default
  - setvalue: odk-instance-first-load
    - any question type with a dynamic default
- output to model:
  - call path: Survey/xml/xml_model/xml_descendent_bindings -> EntityDeclaration/xml_bindings/_get_setvalue_node_for_id
  - setvalue: odk-instance-first-load
    - entities
- output to model:
  - call path: Survey/xml/xml_model/xml_actions -> Question/xml_action
  - odk:setgeopoint: odk-instance-first-load
    - question type start-geopoint
  - odk:recordaudio: odk-instance-load
    - question type background-audio
- output to body/repeat:
  - call path: Survey/xml/xml_control -> RepeatingSection/xml_control/dynamic_defaults_helper -> Question/get_setvalue_node_for_dynamic_default
  - setvalue: odk-instance-first-load, odk-new-repeat
    - any question type with a dynamic default, where the question is in a repeat
- output to body/question control:
  - call path: Survey/xml/xml_control -> Question/xml_control
  - setvalue: xforms-value-changed
    - any question with a trigger
  - odk:setgeopoint: xforms-value-changed
    - background-geopoint

New output locations, call paths, action/event types, reasons for generation

- output to model:
  - call path: Survey/xml/xml_model/xml_model_bindings -> Question/xml_action or EntityDeclaration/xml_action
  - odk:setgeopoint: odk-instance-first-load
    - question type start-geopoint
  - odk:recordaudio: odk-instance-load
    - question type background-audio
  - setvalue: odk-instance-first-load
    - any question type with a dynamic default
    - entities
- output to body/repeat:
  - call path: Survey/xml/xml_control -> RepeatingSection/xml_control -> Question/xml_action
  - setvalue: odk-new-repeat
    - any question type with a dynamic default, where the question is in a repeat
- output to body/question control:
  - call path: Survey/xml/xml_control -> Question/xml_control
  - setvalue: xforms-value-changed
    - any question with a trigger
  - odk:setgeopoint: xforms-value-changed
    - background-geopoint

**Comment by lindsay-stevens** [comment]
> It looks like all of the references produced in binds are off by one parenting level

@lognaturel in 026bfe0 I've fixed this for label - see commit message for details. The paths for the other attribute bindings seemed like they should be OK, assuming that `current()` resolves to the `entity` element - but if that is not the case and the attribute binds also need to go up one level please confirm.

For example from `test_update_repeat.py` test `test_all_fields__ok`, this is the repeat output XML fragment:

```
          <r1>
            <q1/>
            <meta>
              <entity dataset="e1" id="" update="1" baseVersion="" trunkVersion="" branchId="" create="1">
                <label/>
              </entity>
            </meta>
          </r1>
```

And there's a `__version` binding as follows so it'd be current() = `entity` and 2 steps up through `r1/meta` to find `q1`:

```
      <bind
          nodeset="/test_name/r1/meta/entity/@baseVersion"
          calculate="instance('e1')/root/item[name= current()/../../q1 ]/__version"
          type="string" readonly="true()"/>
```

**Comment by lognaturel** [comment]
`current()` definition: https://www.w3.org/TR/xslt-10/#function-current
binding expressions in XForms: https://www.w3.org/TR/2001/WD-xforms-20010608/slice6.html#expr-canonical

If `current()` behaved differently for element nodes and attribute nodes, I think one of those resources would call it out. As the definition for `current()` states, the intent is for `.` and `current()` to evaluate to the same context node when in an expression that's not in a predicate. `.` in an expression bound to an attribute evaluates to that attribute node so `current()` should as well. Does that sound right?

File: pyxform/entities/entities_parsing.py [hunk 4]
```diff
@@ -87,18 +87,30 @@ def get_entity_declaration(
...
             EC.DATASET.value: dataset_name,
             EC.ENTITY_ID.value: entity_id,
             EC.CREATE_IF.value: create_condition,
             EC.UPDATE_IF.value: update_condition,
-            EC.LABEL.value: entity_label,
             EC.REPEAT.value: entity_repeat,
         },
     }
+    if entity_label:
+        entity[const.CHILDREN] = [
```
Comment (lognaturel) [inline pyxform/entities/entities_parsing.py:102]: Ah, I see, that's a nice solution.

**Comment by lognaturel** [comment]
It's feeling really close! I think the only must-have is getting clarity on `current()` in an expression bound to an attribute. I'm pretty confident in my reasoning above but would like your careful scrutiny and possibly some more brains on it. Erroring when `saveto`s are not in scope could come in a follow-on, possibly along with some thinking about multiple Entity declarations in a form def.

> I found actions generation quite confusing

You are not alone. They've been conceptually tricky to fit in everywhere in the ODK XForms world and because we've introduced different actions and events over time the implementations have all gotten messy.

Did I understand correctly that dynamic defaults in repeats were correctly generating `odk-new-repeat`-triggered actions in the body and that the issue was actually just with the the new Entity-related actions? If that's right, it might make sense to move forward with https://github.com/getodk/javarosa/pull/829. Prior to this PR, were there ever `odk-new-repeat`-triggered actions generated in the model?

> I couldn't see any other reason besides brevity to combine the setvalue for that case

I can't either so it looks fine to me. As far as I know, any client implementation would need to separate out the actions by triggering event anyway so it makes no difference on the client side.

> a typo that the entity setvalue had until now included the attributes readonly="true()" and type="string"

Oh yes, good catch! I feel like my brain tripped up on that a few times but not enough to notice it didn't make sense.

**Comment by lognaturel** [comment]
Here's a way to confirm that the "current" node for an attribute is the attribute itself, not its parent:
1. Go to https://xslttest.appspot.com/
1. Specify the following XML:
```
<r1>
  <q1/>
  <meta>
    <entity dataset="e1" id="" update="1" baseVersion="foo" trunkVersion="" branchId="" create="1">
      <label/>
    </entity>
  </meta>
</r1>
```
3. Specify the following XSLT:
```

<?xml version="1.0" encoding="UTF-8"?>
<xsl:stylesheet version="1.0"
xmlns:xsl="http://www.w3.org/1999/XSL/Transform">

<xsl:template match="/">
  <html>
  <body>
  <xsl:for-each select="r1/meta/entity/@baseVersion">
    Current node: <xsl:value-of select="current()"/>
    <br />
  </xsl:for-each>
  </body>
  </html>
</xsl:template>

</xsl:stylesheet> 
```
4. Confirm that the output is 
```
<html>
   <body>
      Current node: foo<br></body>
</html>
```

**Comment by lindsay-stevens** [comment]
Yes it seems now like a pretty foundational concept; I didn't consider attributes as a "node" as such. But the spec says everything is a node, and 0014ca9 is a moving towards everything is a SurveyElement. I still think the `Element` class idea added in 1e481b9c for actions could be useful though, for defining the [type / supported attributes / values] of generated nodes e.g. `bind`, `type`, `control`, etc.

https://www.w3.org/TR/xforms11/%23context-xpath-evaluation#fn-current

> 7.10.2 The current() Function
> 
> node-set current()
> 
> Returns the context node used to initialize the evaluation of the containing XPath expression.

https://www.w3.org/TR/1999/REC-xpath-19991116/#section-Introduction

> XPath models an XML document as a tree of nodes. There are different types of nodes, including element nodes, attribute nodes and text nodes. 

https://www.w3.org/TR/1999/REC-xpath-19991116/#data-model

> There are seven types of node:
> - root nodes
> - element nodes
> - text nodes
> - attribute nodes
> - namespace nodes
> - processing instruction nodes
> - comment nodes
